Enable implicit fallthrough warning

janvorli opened the pull request:
This change enables warnings for implicit fallthrough in
switch cases and fixes all the cases where the warning
was reported.

Close #43308

janvorli committed:
Enable implicit fallthrough warning

This change enables warnings for implicit fallthrough in
switch cases and fixes all the cases where the warning
was reported.

janvorli commented:
The first commit fixes all coreclr issues. I've found a couple of places that look like fallthrough bugs at various places, in some cases I was not sure yet.
@dotnet/jit-contrib could someone please look at the fixes in the JIT?

janvorli commented:
I've actually tested it with x64 build only yet, I am just about to try arm and arm64 too.
I also need to check / fix installer and libraries.

sandreenko commented on src/coreclr/src/jit/lower.cpp:
```diff
@@ -6278,6 +6278,7 @@ void Lowering::ContainCheckNode(GenTree* node)
             break;
         case GT_STOREIND:
             ContainCheckStoreIndir(node->AsIndir());
+            __fallthrough; // TODO: is this actually a bug?
```
Please change it to `break`, it does not lead to an incorrect execution, but `ContainCheckStoreIndir` calls `ContainCheckIndir` itself and we don't need to call it again (but it does not hurt correctness).

sandreenko commented on src/coreclr/src/jit/lowerxarch.cpp:
```diff
@@ -5594,6 +5594,7 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                             }
                         }
                     }
+                    break;
```
can we use `unreached(); break;` in such cases?

janvorli replied:
@sandreenko in the CI run, 15 coreclr tests have failed due to this added unreachable. So it seems it should be just a break here. I guess the fall through that was here before was benign, as the switch in the case below would end up in the default and nothing wrong would happen.

janvorli committed:
Fix ARM build

janvorli committed:
Fix libraries build

janvorli committed:
Fix OSX

janvorli commented:
I've added commits fixing ARM/ARM64 coreclr and libraries build and also OSX x64 build.

janvorli committed:
Reflect PR feedback

janvorli commented:
Hmm, the libraries build is failing, since the C `__attribute__((fallthrough));` is not supported by clang 9. My local builds pick clang 10 which already supports it. It seems I'll need to define a macro like we have for coreclr and make it empty for clang < 10 (and gcc < 7).

janvorli committed:
Fix issue with clang < 10

Clang version < 10 doesn't support the __attribute__((fallthrough))
in C code.

janvorli committed:
Minor fix

janvorli committed:
Fix build with clang 9

It turned out that clang 9 doesn't support the __attribute__((fallthrough))
in C++, only in C. So my previous change that unified the attribute
to this one in both languages didn't work on clang 9 (was working
fine on clang 10)

janvorli committed:
Remove incorrectly added unreachable

janvorli committed:
Apply formatting patch

janvorli committed:
Cleanup

janvorli commented on src/coreclr/src/jit/optimizer.cpp:
```diff
@@ -7561,6 +7561,7 @@ void Compiler::fgCreateLoopPreHeader(unsigned lnum)
                         preHead->bbFlags |= BBF_JMP_TARGET | BBF_HAS_LABEL;
                     }
                 } while (++jumpTab, --jumpCnt);
+                __fallthrough; // TODO: seems like a bug
```
@sandreenko I guess this should be break instead, otherwise the  BBJ_SWITCH always ends up in the noway_assert below, right?

janvorli replied:
Ah, I can see @sandreenko is OOF, @dotnet/jit-contrib can someone else please answer my question?

sandreenko replied:
Yeah, it looks like it should be `break` but currently this code is unreachable, I will check why.

janvorli commented on src/coreclr/src/ildasm/dasm_sz.cpp:
```diff
@@ -211,6 +211,7 @@ unsigned SizeOfField(PCCOR_SIGNATURE *ppSig, ULONG cSig, IMDInternalImport* pImp
                     case ELEMENT_TYPE_CMOD_OPT  :
                     case ELEMENT_TYPE_CMOD_REQD :
                         *ppSig += CorSigUncompressToken(*ppSig, &tk);
+                         __fallthrough; // TODO: is this correct?
```
@davidwrighton I am not sure if this fallthrough here is intentional, can you please advise me on this one?

davidwrighton replied:
Yes, this looks intentional. I don't think its well written, but intentional, and not functionally wrong... yes.

davidwrighton commented:
Could you enable this for msvc builds as well? Its astonishing annoying to have major differences in warning behavior between windows and linux. Also, if possible I'd prefer to rely on the [[falthrough]]; statement which as a standard part of C++ 17 should be supported on both recent msvc and sufficiently recent clang.

The warning is C26819 for msvc.

janvorli commented:
@davidwrighton this change uses the `[[fallthrough]]` attribute already on Linux except of C code (libraries) that has to use the `__attribute__((fallthrough))`  as clang doesn't support the `[[fallthrough]]` in C code. 
As for enabling this for Windows, I am a bit hesitant to move to C++17 just because of this. Currently we set the compiler to use C++11 on both Windows and Unix. @jkotas any thoughts on this?

jkotas commented on src/coreclr/src/gc/unix/gcenv.unix.cpp:
```diff
@@ -28,6 +28,12 @@
 #undef max
 
+#if __has_cpp_attribute(fallthrough)
+#define FALLTHROUGH [[fallthrough]];
```
Can this be `#define __fallthrough [[fallthrough]]` ?

janvorli replied:
Symbols starting with underscore are reserved for the compiler and the compiler warns. We have the warning disabled for coreclr, but not for PAL. Disabling it for coreclr would be complicated due to all of the SAL stuff. But maybe we should move coreclr to use FALLTHROUGH too so that it is uniform.

janvorli replied:
Actually - a small correction - I didn't hit the warning in PAL, but in libraries minipal. I had __fallthrough in PAL before and then I've unified PAL and libraries minipal on the uppercase version.

janvorli replied:
I also tried to use the attribute directly without the macro, but that broke GCC build as it was not supported until GCC 7

jkotas replied:
Yes, it would be nice to use same macro everywhere. I see number of other projects using `#define FALLTHROUGH` : https://github.com/search?p=2&q=%22%23define+fallthrough%22&type=Code.

jkotas commented on src/libraries/Native/Unix/Common/pal_utilities.h:
```diff
@@ -45,6 +45,12 @@
 #define ARRAY_SIZE(a) (sizeof(a)/sizeof(a[0]))
 
+#if __has_attribute(fallthrough)
+#define FALLTHROUGH __attribute__((fallthrough));
```
Same here

jkotas replied:
```suggestion
#define FALLTHROUGH __attribute__((fallthrough))
```

jkotas commented:
> hesitant to move to C++17

Agree.

davidwrighton commented:
> > hesitant to move to C++17
> 
> Agree.

Sadly, yes, I agree this isn't quite enough of a reason to move to C++17. I would like to move eventually though, as the current compilation language for msvc is actually C++14 + some amount of C++17 features that had shipped before the msvc team decided to add C++ language version flags to the compiler. When we move to C++17 we will actually enter into a config where we can specify that both msvc and clang should attempt to implement the same C++ language, and if that transition happens to bring somewhat useful tweaks like this along, I'm all for it.

janvorli commented:
I believe we use C++11 for coreclr and not C++14 on Windows too. We have it set here for both clang and MSVC:
https://github.com/dotnet/runtime/blob/7601d44ca06fc977c4a3c7a33c1fc864f8f5ad78/eng/native/configurecompiler.cmake#L6-L8

davidwrighton commented:
Yes that is what is specified in cmake; however, cl.exe doesn't have a C++11 mode. It does have a `/std:c++14` switch, but that switch isn't limited to C++ 14 either, as some feature of C++17 had been implemented before the standard specifying switch was implemented. If you run `cl.exe /?` in a developer command prompt, the available switches will be displayed.

davidwrighton commented:
See https://devblogs.microsoft.com/cppblog/standards-version-switches-in-the-compiler/ for details.

janvorli commented:
Interesting, thank you for the details!

jkotas commented on src/libraries/Native/AnyOS/brotli/dec/decode.c:
```diff
@@ -24,6 +24,16 @@
 #include <arm_neon.h>
 #endif
 
+#ifndef __has_attribute
```
We should disable warning for brotli instead. We typically avoid fixing warnings in the vendored code.

janvorli replied:
Makes sense.

jkotas commented on src/coreclr/src/gc/unix/gcenv.unix.cpp:
```diff
@@ -28,6 +28,12 @@
 #undef max
 
+#if __has_cpp_attribute(fallthrough)
+#define FALLTHROUGH [[fallthrough]];
```
```suggestion
#define FALLTHROUGH [[fallthrough]]
```

janvorli committed:
Reflect PR feedback

* Disable the fallthrough warning for Brotli sources
* Remove semicolon from the FALLTHROUGH macro definition

janvorli committed:
Replace __fallthrough by FALLTHROUGH

janvorli commented:
@jkotas I've made the changes you've suggested and changed `__fallthrough` to `FALLTHROUGH` in the whole runtime.

janvorli committed:
Add break to Compiler::fgCreateLoopPreHeade switch
